src/reader.rs

```diff
@@ -0,0 +1,939 @@
+                }
+            }
+
+            let name = GCNO::read_string(reader)?;
+            let file_name = GCNO::read_string(reader)?;
+            let line_number = reader.read_u32::<LittleEndian>()?;
+            let block_tag = reader.read_u32::<LittleEndian>()?;
+
+            if block_tag == 0x0141_0000 {
+                let count = reader.read_u32::<LittleEndian>()? as usize;
+                let mut blocks: Vec<GcovBlock> = Vec::with_capacity(count);
+                for no in 0..count {
+                    blocks.push(GcovBlock {
+                        no,
+                        flags: reader.read_u32::<LittleEndian>()?,
+                        source: Vec::new(),
+                        destination: Vec::new(),
+                        lines: Vec::new(),
+                        line_max: 0,
+                        counter: 0,
+                    });
+                }
+                let mut fun = GcovFunction {
+                    identifier,
+                    line_number,
+                    line_checksum,
+                    file_name,
+                    name,
+                    blocks,
+                    edges: Vec::new(),
+                    lines: HashMap::new(),
+                    executed: false,
+                };
+                tag = GCNO::read_edges(&mut fun, reader)?;
+                tag = GCNO::read_lines(&mut fun, reader, tag)?;
+                self.functions.push(fun);
+            } else {
+                return Err(GcovError::Str(format!("Invalid function tag: {:?}", tag)));
+            }
+        }
+        Ok(())
+    }
+
+    fn read_gcda(&mut self, reader: &mut Read) -> Result<(), GcovError> {
+        GCNO::check_type(reader, [b'a', b'd', b'c', b'g'])?;
+        let version = GCNO::read_version(reader)?;
+        if version != self.version {
+            Err(GcovError::Str("GCOV versions do not match.".to_string()))
+        } else {
+            let checksum = reader.read_u32::<LittleEndian>()?;
+            if checksum != self.checksum {
+                Err(GcovError::Str(format!("File checksums do not match: {:?} != {:?}", self.checksum, checksum)))
+            } else {
+                for mut fun in &mut self.functions {
+                    GCNO::read_gcda_function(self.version, self.checksum, &mut fun, reader)?;
+                }
+                let object_tag = reader.read_u32::<LittleEndian>()?;
+                if object_tag == 0xa100_0000 {
+                    reader.read_u32::<LittleEndian>()?;
+                    reader.read_u64::<LittleEndian>()?;
+                    self.runcounts += reader.read_u32::<LittleEndian>()?;
+                }
+                Ok(())
+            }
+        }
+    }
+
+    fn read_gcda_function(version: u32, checksum: u32, fun: &mut GcovFunction, reader: &mut Read) -> Result<(), GcovError> {
```

Comment:
Try to return the `Err`s early in this function, so you can reduce the amount of indentation.